api/tests/integration/EntityTest.php

```diff
@@ -152,6 +152,92 @@ public function testGetAllPaginatedSimpleRange()
         $this->assertEquals($last, 19);
     }
 
+    public function testGetAllPaginatedSimpleSearch()
+    {
+        $resultsMock = Mockery::mock('Elasticquent\ElasticquentResultCollection');
+        $resultsMock->shouldReceive('totalHits')
+            ->andReturn(0); // Force not found, we don't have to mock a success, just that 'searchByQuery' is called with the right params.
+
+        $mockModel = Mockery::mock(TestEntity::class);
+        $mockModel
+            ->shouldReceive('count')
+            ->andReturn(10)
+            ->shouldReceive('searchByQuery')
+            ->with([
+                'match_phrase' => [
+                    '_all' => 'foobar',
+                ],
+            ], null, null, 10, 0)
+            ->andReturn($resultsMock);
+
+        $this->app->instance(TestEntity::class, $mockModel);
+
+        $this->getJson('/test/entities/pages?q=foobar', ['Range' => 'entities=0-']);
+
+        $this->assertResponseStatus(404);
+    }
+
+    public function testGetAllPaginatedComplexSearch()
+    {
+        $resultsMock = Mockery::mock('Elasticquent\ElasticquentResultCollection');
+        $resultsMock->shouldReceive('totalHits')
+            ->andReturn(0); // Force not found, we don't have to mock a success, just that 'searchByQuery' is called with the right params.
+
+        $mockModel = Mockery::mock(TestEntity::class);
+        $mockModel
+            ->shouldReceive('count')
+            ->andReturn(10)
+            ->shouldReceive('getIndexName')
+            ->andReturn('defaultIndex')
+            ->shouldReceive('getTypeName')
+            ->andReturn('someTypeName')
+            ->shouldReceive('complexSearch')
+            ->with([
+                'index' => 'defaultIndex',
+                'type' => 'someTypeName',
+                'body' => ['query' => ['bool' => ['must' => [['match' => ['_all' => 'search term']],['match' => ['author_id' => 'some UUID']],['nested' => ['path' => 'tags','query' => ['bool' => ['must' => ['match' => ['tags.tag_id' => 'tag ID 1']]]]]],['nested' => ['path' => 'tags','query' => ['bool' => ['must' => ['match' => ['tags.tag_id' => 'tag ID 2']]]]]]]]]],
+            ])
+            ->andReturn($resultsMock);
+
+        $this->app->instance(TestEntity::class, $mockModel);
+
+        $this->getJson('/test/entities/pages?q=%7B%22_all%22%3A%5B%22search%20term%22%5D%2C%22authorId%22%3A%5B%22some%20UUID%22%5D%2C%22_tags%22%3A%7B%22tagId%22%3A%5B%22tag%20ID%201%22%2C%20%22tag%20ID%202%22%5D%7D%7D', ['Range' => 'entities=0-']);
```

Comment:
We should base64 encode our json strings so that we don't have awkward url encoding (this can lead to unexpected substitutions, base64 is always url safe)